### PR TITLE
Bug Fix: Allow Unique Field Merge in Contact Merge

### LINKED ISSUE #6161: NPSP Contact Merge Error When Unique Field Value Chosen From Losing Contact Record
When a customer selects a value from a "losing" (aka non-master) Contact's field marked as unique and attempts to merge the Contacts using the NPSP Contact Merge tool, an error prevents the merge:
_Error: Failed to merge the contacts. Error: duplicate value found: My_Custom_External_Id__c duplicates value on record with id: 003xxxxxxxxxxxx”_
**Note**: This is related to Issue #1759, but likely closed since we didn't have exact test conditions. 

Screen Shot 2021-01-08 at 3 51 22 PM

### Root Cause
[In CON_ContactMerge_CTRL's mergeContacts() method,](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/CON_ContactMerge_CTRL.cls#L940-L1005) we update the Master record first. Since the losing Contact record already contains that unique Id, the error is thrown.

### Steps to Repeat
1. Create a new Contact field called “My Custom External Id”. Mark it as Unique.
2. Create a Contact (First Name = Test; Last Name, = Test; My Custom External Id = abc123)
3. Create a second Contact (First Name = Test; Last Name, = Two; My Custom External Id = def234)
4. Go to the Contact Merge Tab and click the Search Contacts button. Enter “test” in the search box and hit the Search button.
5. Select both Contacts and hit Next. 
6. On the record that’s not the Master Record (there is a field near the top indicating which record is the Master Record), select the radio button for the My Custom External Id field (i.e. select the “losing” Contact’s field value).
7. Click Merge and Merge again in the modal.
8. Confirm error appears: “Error: Failed to merge the contacts. Error: duplicate value found: My_Custom_External_Id__c duplicates value on record with id: 003xxxxxxxxxxxx”

### Other Testing Notes
* The error does not appear if you don't mark the field as Unique. 
* The error also appears if you choose External Id AND Unique.
* Standard Salesforce Contact Merge does not fail.

### PR DESCRIPTION
**Pull request by andrewyu-salesforce**
[W-8668236](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008w3EyIAI/view)
# Critical Changes

# Changes

# Issues Closed
Fixes #6161
# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (lparrott) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:975]: Are we potentially introducing a regression by changing this?

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (lparrott) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1014]: Small typo, should be `losingContacts` I think?

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (lparrott) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1045]: For now, I think a Unit test that checks for our "test" unique field on Contact could be sufficient to cover this method. We can create a separate test or flow later.

**Review by lparrott — COMMENTED** [review]
Just adding a few comments after a quick review, I know this is still in progress!

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (andrewyu-salesforce) [reply force-app/main/default/classes/CON_ContactMerge_CTRL.cls:975]: I dont think so, I was testing this with apex anonymous and it seem like the put can accept null value now. I can roll back this change if needed.

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (andrewyu-salesforce) [reply force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1045]: I add a whole regression flow to this

File: cumulusci.yml
Comment (andrewyu-salesforce) [inline cumulusci.yml:887]: Add a new flow to deploy unpackaged contact merge testing.

File: unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls
Comment (andrewyu-salesforce) [inline unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls:2]: I use `REG__` instead of `CON_` as prefix on the test so in the future, regression flow can query other tests to run

File: cumulusci.yml
Comment (andrewyu-salesforce) [inline cumulusci.yml:1184]: a new flow to deploy contact merge testing and potentially all other tests that requires metadata that does not include in the package.

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (lparrott) [reply force-app/main/default/classes/CON_ContactMerge_CTRL.cls:975]: I'll do some testing and let you know!

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (lparrott) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1070]: This is a great refactor 👍

File: force-app/main/default/classes/UTIL_Describe.cls
Comment (lparrott) [inline force-app/main/default/classes/UTIL_Describe.cls:476]: This is a great addition to our Describe class.

File: unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls
Comment (lparrott) [reply unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls:2]: Works for me! Also, we can easily change these later since they're unpackaged.

**Review by lparrott — COMMENTED** [review]
Code looks great, I'm going to switch over to more testing before I officially approve.

**Review by lparrott — APPROVED** [review]
My testing has all come back positive, great work on this!

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (lparrott) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:989]: @andrewyu-salesforce Would it be possible to only call this if a field was changed? Merging Unique fields will be pretty uncommon...

**Comment by dfuller114** [comment]
@andrewyu-salesforce should we include someone from RelEng to review the changes to the .yaml file?

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (andrewyu-salesforce) [reply force-app/main/default/classes/CON_ContactMerge_CTRL.cls]: The boolean check is added

File: force-app/main/default/classes/CON_ContactMerge_CTRL.cls
Comment (dfuller114) [inline force-app/main/default/classes/CON_ContactMerge_CTRL.cls:1038]: it looks like there's duplicate code where with the same code block in `populateContactWithOnlyUniqueFields()`. Could we extract this logic to a method to call in both places?

File: unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls
Comment (dfuller114) [inline unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls:90]: Since `mergeContacts()` is wrapped in a try/catch in the controller, should we add an assert here to confirm that no exception has been thrown with the unique field?

**Review by dfuller114 — CHANGES_REQUESTED** [review]
great work Andrew! just had a few commments

File: unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls
Comment (andrewyu-salesforce) [reply unpackaged/config/con_merge_testing/classes/REG_ContactMergeUniqueField_TEST.cls:90]: add new assert to check if an error is thrown onto the page

**Review by jstvz — APPROVED** [review]
CODEOWNER 🆗

**Review by dfuller114 — APPROVED** [review]
looks good!

**Pull request merged**
